### PR TITLE
feat: Allow Kustomize common labels to not apply to selectors

### PR DESCRIPTION
**Pull request by lets-call-n-walk**
* Closes issue https://github.com/argoproj/argo-cd/issues/17328
* Argo commonlabels in an application's kustomize source overrides selectors and templates. Should have a way to use the kustomize feature `--without-selector` for common labels.
PS, sorry if I'm doing something wrong, this is my first time contributing to any open source project.

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [x] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [x] Does this PR require documentation updates?
* [x] I've updated documentation as required by this PR.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [x] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [x] I have added a brief description of why this PR is necessary and/or what this PR solves.
* [x] Optional. My organization is added to USERS.md.
* [x] Optional. For bug fixes, I've indicated what older releases this fix should be cherry-picked into (this may or may not happen depending on risk/complexity).

### CODE REVIEW

**Review by ishitasequeira — CHANGES_REQUESTED** [review]

**Review by ishitasequeira — CHANGES_REQUESTED** [review]
@lets-call-n-walk Please check my comment and also you would need to execute `make codegen` or `make codegen-local` to update the generated code.

**Comment by lets-call-n-walk** [comment]
@ishitasequeira Thank you for your guidance. I have tried running `make codegen` and `make codegen-local`. The docker image gets to a step
```
ERROR: failed to solve: executor failed running [/bin/sh -c ./install.sh helm-linux &&     ./install.sh kustomize &&     ./install.sh codegen-tools &&     ./install.sh codegen-go-tools &&     ./install.sh lint-tools &&     go install github.com/mattn/goreman@latest &&     go install github.com/kisielk/godepgraph@latest &&     go install github.com/jstemmer/go-junit-report@latest &&     rm -rf /tmp/dl &&     rm -rf /tmp/helm &&     rm -rf /tmp/ks_*]: exit code: 60
make: *** [test-tools-image] Error 1
```
make codegen-local gives the following output. I can't seem to find a way to install this particular piece of protobuf through `brew`
```
./hack/generate-proto.sh: line 60: go-to-protobuf: command not found
```

**Comment by blakepettersson** [comment]
For `make codegen-local` to work, you first need to run `make install-tools-local`

**Comment by lets-call-n-walk** [comment]
@blakepettersson thanks

**Review by ishitasequeira — CHANGES_REQUESTED** [review]
@lets-call-n-walk it seems the commit history got messed up. Can you clean it up? The PR shows way more changes than needed because of it.

**Review by ishitasequeira — CHANGES_REQUESTED** [review]
You would also need to set the value of the flag from the switch case like https://github.com/argoproj/argo-cd/blob/11210fc9b820d126889bed4db55f0fe6af2848cf/cmd/util/app.go#L233-L236

**Comment by lets-call-n-walk** [comment]
@ishitasequeira I think I've cleaned it up. I think I messed it up initially with my rebase in order to sign off the commits. After a few failed attempts, I think I have everything how it should be. This has taught me alot more about git.

**Comment by lets-call-n-walk** [comment]
@ishitasequeira can you see why the e2e 1.25 test failed? Seems like health check timed out for a config map, but I can't tell how that is related.

**Comment by lets-call-n-walk** [comment]
Looks like updating the branch fixed whatever was going on.

**Review by jannfis — COMMENTED** [review]
This is great! Thank you @lets-call-n-walk 

I have a question, why is there a "beta" in the PR's title? Is it a beta feature in Kustomize?

**Comment by lets-call-n-walk** [comment]
@jannfis no, I just thought I had to put something about the status of the feature according to the PR task list. If I misunderstood, we can change that to whatever it needs to be.

File: util/kustomize/kustomize_test.go [hunk 1]
```diff
@@ -328,6 +329,87 @@ func TestKustomizeBuildForceCommonAnnotations(t *testing.T) {
...
+				&v1alpha1.EnvEntry{
+					Name:  "ARGOCD_APP_NAME",
+					Value: "argo-cd-tests",
+				},
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		fmt.Println("beginning")
```
Comment (jannfis) [inline util/kustomize/kustomize_test.go:383]: Is this a debugging leftover?

File: util/kustomize/kustomize_test.go [hunk 2]
```diff
@@ -328,6 +329,87 @@ func TestKustomizeBuildForceCommonAnnotations(t *testing.T) {
...
+		appPath, err := testDataDir(t, tc.TestData)
+		assert.Nil(t, err)
+		fmt.Println(appPath)
```
Comment (jannfis) [inline util/kustomize/kustomize_test.go:386]: Is this a debugging leftover?

File: util/kustomize/kustomize_test.go [hunk 3]
```diff
@@ -328,6 +329,87 @@ func TestKustomizeBuildForceCommonAnnotations(t *testing.T) {
...
+		kustomize := NewKustomizeApp(appPath, appPath, git.NopCreds{}, "", "")
+		objs, _, err := kustomize.Build(&tc.KustomizeSource, nil, tc.Env)
+		fmt.Println("objs: ", len(objs))
```
Comment (jannfis) [inline util/kustomize/kustomize_test.go:389]: Is this a debugging leftover?

File: util/kustomize/kustomize_test.go [hunk 4]
```diff
@@ -328,6 +329,87 @@ func TestKustomizeBuildForceCommonAnnotations(t *testing.T) {
...
+		switch tc.ExpectErr {
+		case true:
+			assert.Error(t, err)
+		default:
+			assert.Nil(t, err)
+			if assert.Equal(t, len(objs), 1) {
+				fmt.Println(len(objs))
```
Comment (jannfis) [inline util/kustomize/kustomize_test.go:396]: Is this a debugging leftover?

**Comment by lets-call-n-walk** [comment]
@jannfis Thanks for the catch. Removed.

There seems to be a conflict with Master now where a bunch of app options in the cli are missing... Is this because of some major change? What is the best way to proceed?

**Comment by jannfis** [comment]
@lets-call-n-walk Yes, we merged [another PR](https://github.com/argoproj/argo-cd/pull/17360) that touched the same part of the code on Friday and moved some things around, so your change can't be merged without conflict. You need to rebase your PR and see how to resolve the conflict.

**Comment by lets-call-n-walk** [comment]
@jannfis Hmm, I don't see where that PR touches `cmd/util/app.go`. But I will hunt around for where the kustomize opts got moved to.

**Comment by jannfis** [comment]
@lets-call-n-walk, I have unfortunately linked the wrong PR, sorry about that.

It was of course [this one](https://github.com/argoproj/argo-cd/pull/17310).

**Comment by lets-call-n-walk** [comment]
@jannfis Ok, I think I fixed it. Thank you!

**Review by jannfis — APPROVED** [review]
LGTM, thank you! Well done, and congratulations to your first contribution @lets-call-n-walk :)

**Comment by jannfis** [comment]
@ishitasequeira Could you please check status of your RfC here?

**Comment by lets-call-n-walk** [comment]
@jannfis Thank you!  Just curious, do you have an idea of what release this will be packaged in, and very roughly, when we can expect to see that release available?

**Comment by jannfis** [comment]
@lets-call-n-walk This will be part of the v2.11 release. We do have a policy for not cherry-picking new features into previous releases. 2.11 is planned for May.

**Review by ishitasequeira — APPROVED** [review]

**Comment by omer2500** [comment]
Hi,

can you explain @lets-call-n-walk why didn't you choose to implement this feature:
https://kubectl.docs.kubernetes.io/references/kustomize/kustomization/labels/

it seems that it's the modern way to go with kustomize and already have all the flags that cover all the use cases
and commonLabels is already deprecated in kustomize 

don't we want to support the labels to allow all abilities?

#15283

**Comment by lets-call-n-walk** [comment]
@omer2500 the short answer is that Argo builds applications via the kustomize cli, and the kustomize cli has not fully reached parity with the new kustomize labels API.

The only way to add labels via the cli is to use the flag `--commonlabels`, which obviously adds the labels as `commonLabels`. Thankfully, they have added a flag `--label-without-selector`. Unexpectedly, it makes `--commonlabels` implement `labels` and add the `includeSelectors: false` field. No further development has occurred to my knowledge.

Because the Argo kustomize application source is built around the kustomize cli, I thought it prudent to keep the same fields.

**Comment by omer2500** [comment]
Thanks for the detailed answer! @lets-call-n-walk

**Pull request merged**